near-contract-standards/src/access/mod.rs

```diff
@@ -1,129 +1,279 @@
+                }
+            }
+        }
     }
 
-    pub fn revoke_role(&mut self, role: [u8; 32], account: AccountId) {
+    /// Revokes [`role` from `account`](RoleData::members).
+    ///
+    /// Requirements:
+    ///
+    /// - The [`role`](Self::roles) must exist.
+    /// - The caller must have the `role`'s [admin role](RoleData::admin_role).
+    ///
+    /// Uses [`Self::only_role()`] internally.  
+    ///
+    /// See also [`Self::grant_role()`] for the opposite effect.
+    pub fn revoke_role(&mut self, role: RoleId, account: AccountId) {
         self.only_role(&self.get_role_admin(&role));
-        if self.has_role(&role, &account) {
-            self.roles.get_mut(&role).unwrap().members.insert(account, false);
+        self.internal_revoke_role(role, account);
+    }
+
+    /// Revokes [`role` from `account`](RoleData::members).
+    ///
+    /// Has no effect if the [`role`](Self::roles) does not exist,
+    /// or if the [`account`](RoleData::members) is not a member for that `role`,
+    /// or if the `account` was already disabled.
+    fn internal_revoke_role(&mut self, role: RoleId, account: AccountId) {
+        use std::collections::hash_map::Entry;
+        match self.roles.entry(role) {
+            // `role` exists
+            Entry::Occupied(mut role_entry) => {
+                match role_entry.get_mut().members.entry(account) {
+                    // `account` is a member, update status
+                    Entry::Occupied(mut member_entry) => {
+                        *member_entry.get_mut() = false;
+                    }
+
+                    // `account` isn't a member, no action needed
+                    Entry::Vacant(_) => {}
+                }
+            }
+
+            // `role` inexistent, no action needed
+            Entry::Vacant(_) => {}
         }
     }
 
-    pub fn renounce_role(&mut self, role: [u8; 32], account: AccountId) {
+    /// Revokes [`role` from](RoleData::members) the [predecessor's `account`](env::predecessor_account_id()).  
+    /// Has no effect if the [`role`](Self::roles) does not exist, or if the [`account`
+    /// is not enabled](RoleData::members) for that `role`.
+    ///
+    /// This method's purpose is to provide a mechanism for `account`s to purposefuly
+    /// [lose their own privileges](RoleData::members), such as when they are compromised
+    /// (eg. when a trusted device is misplaced).
+    ///
+    /// Requirements:
+    ///
+    /// - The `account` must be the [predecessor account](env::predecessor_account_id()).
+    ///
+    /// See also [`Self::grant_role()`] and [`Self::revoke_role()`] for other management of
+    /// roles.
+    pub fn renounce_role(&mut self, role: RoleId, account: AccountId) {
         require!(
             account == env::predecessor_account_id(),
             "AccessControl: can only renounce roles for self"
         );
-        self.revoke_role(role, account);
+        self.internal_revoke_role(role, account);
```

Comment:
This has an effect difference because previously, using revoke_role, it would also verify if the caller is the admin for that role; but with the change, it passes without this verification (as the acc is revoking it's own privileges)